### DISCUSSION TITLE
Will grpc proxy's get request transparently transmit the request to etcd?

### DISCUSSION THREAD

**Discussion by robotLJW** [etcd-io/etcd#16808] [Q&A] [answered]
**Scenes**：

The two clients use grpc proxy to watch the key with the xxx prefix. Now a new client needs to get the value of the /xxx prefix key. 

![image](https://github.com/etcd-io/etcd/assets/27326092/57fbe907-5b40-4092-8b03-31009cc77fc6)

**My question**：

Will the request be passed transparently to etcd? Does gRPC proxy have relevant cache?

Thanks

**Comment by jmhbnz** [comment, marked as answer]
Hey @robotLJW - Thanks for your question, the etcd grpc proxy documentation is available here: https://etcd.io/docs/v3.5/op-guide/grpc_proxy

Relevant sections are:

> To protect the cluster against abusive clients, it caches key range requests.

> The gRPC proxy caches responses for requests when it does not break consistency requirements. This can protect the etcd server from abusive clients in tight for loops.

The key thing appears to be the mention of "**consistency requirements**" in the documentation above. Based on that I believe only serializable requests are cached. Refer to the API reference for context on serializable: https://etcd.io/docs/v3.5/dev-guide/api_reference_v3/#message-rangerequest-apietcdserverpbrpcproto 

> serializable sets the range request to use serializable member-local reads. Range requests are linearizable by default; linearizable requests have higher latency and lower throughput than serializable requests but reflect the current consensus of the cluster. For better performance, in exchange for possible stale reads, a serializable range request is served locally without needing to reach consensus with other nodes in the cluster.


So long story short yes I believe the grpc proxy will cache serializable range requests, however the trade-off for this increased performance is the potential for stale reads.

**Reply by robotLJW** [reply]
@jmhbnz Thank you for your answer. I found the relevant code : https://github.com/etcd-io/etcd/blob/406cb94b6f9880fddcd9cf5619f89415dee68df3/server/proxy/grpcproxy/kv.go#L41.

**Reply by robotLJW** [reply]
@jmhbnz . But now I have a new problem. Basically at the same time, there are 100 **linearizable requests** getting the key prefixed by /xxx. How to deal with it? How to reduce etcd load?
![image](https://github.com/etcd-io/etcd/assets/27326092/791f6ff5-34c2-4154-b8c6-a8e1a199e672)

**Reply by serathius** [reply]
For caching linearizable requests you need https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache. I don't think it's implemented by proxy.

**Reply by robotLJW** [reply]
@serathius thank you for your answer.
